stl/inc/flat_map

```diff
@@ -0,0 +1,873 @@
+            --*this;
+            return _Tmp;
+        }
+
+        type& operator+=(difference_type _Off) {
+            _Key_it += _Off;
+            _Mapped_it += _Off;
+            return *this;
+        }
+
+        type& operator-=(difference_type _Off) {
+            _Key_it -= _Off;
+            _Mapped_it -= _Off;
+            return *this;
+        }
+
+        type operator+(difference_type _Off) const {
+            type _Tmp = *this;
+            return _Tmp += _Off;
+        }
+
+        type operator-(difference_type _Off) const {
+            type _Tmp = *this;
+            return _Tmp -= _Off;
+        }
+
+        reference operator[](difference_type _Off) const {
+            return *(*this + _Off);
+        }
+
+        difference_type operator-(const type& _Right) const {
+            return _Key_it - _Right._Key_it;
+        }
+
+        friend type operator+(difference_type _Off, const type& _Right);
+
+        template <bool _Val = _IsConst, _STD enable_if_t<!_Val, int> = 0>
+        operator typename _Flat_map_iterator_Impl<_KeyContainer, _MappedContainer, true>::type() const {
+            return typename _Flat_map_iterator_Impl<_KeyContainer, _MappedContainer, true>::type{_Key_it, _Mapped_it};
+        }
+
+    private:
+        _Key_iterator_t _Key_it;
+        _Mapped_iterator_t _Mapped_it;
+    };
+
+    static_assert(_STD swappable<type>);
+};
+
+template <class _KeyContainer, class _MappedContainer, bool _IsConst>
+_EXPORT_STD typename _Flat_map_iterator_Impl<_KeyContainer, _MappedContainer, _IsConst>::type operator+(
+    typename _Flat_map_iterator_Impl<_KeyContainer, _MappedContainer, _IsConst>::type::difference_type _Off,
+    const typename _Flat_map_iterator_Impl<_KeyContainer, _MappedContainer, _IsConst>::type& _Right) {
+    return _Right + _Off;
+}
+
+_EXPORT_STD
+template <class Key, class T, class Compare, class KeyContainer, class MappedContainer>
+    requires same_as<Key, typename KeyContainer::value_type> && same_as<T, typename MappedContainer::value_type>
+class flat_map {
+public:
+    using key_type              = Key;
+    using mapped_type           = T;
+    using value_type            = pair<key_type, mapped_type>;
+    using key_compare           = Compare;
+    using reference             = pair<const key_type&, mapped_type&>;
+    using const_reference       = pair<const key_type&, const mapped_type&>;
+    using size_type             = size_t;
+    using difference_type       = ptrdiff_t;
+    using key_container_type    = KeyContainer;
+    using mapped_container_type = MappedContainer;
+    using iterator = typename _STD _Flat_map_iterator_Impl<key_container_type, mapped_container_type, false>::type;
+    using const_iterator = typename _STD _Flat_map_iterator_Impl<key_container_type, mapped_container_type, true>::type;
+    using reverse_iterator       = _STD reverse_iterator<iterator>;
+    using const_reverse_iterator = _STD reverse_iterator<const_iterator>;

```

Comment:
The `_STD` looks useful here.

Reply:
I added them everywhere because I couldn't figure out how to make things ADL-proof and I got an impression that it's always better to spell out `::std::` 😂